code refactor

nemerna opened the pull request:
1) replaced the authentication method to the one requested by Babak in the ticket

2) removed the central control script

3) containerized the application

4) added Makefile for make easy run

5) reduced the number of needed parameters

nemerna committed:
code refactor

nemerna commented:
@ilan-pinto

ilan-pinto commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+FROM quay.io/ecosystem-appeng/keycloak-source:latest as keycloak
+
+FROM registry.access.redhat.com/ubi9/openjdk-11-runtime
```
Why did you pick this image? 
i don't think it's the latest

nemerna replied:
i did not chose it. i Reused from 
https://github.com/dmartinol/keycloak-change-admin-kcadm/tree/main/docker

youe mean the openjdk or the Keycloak one?

ilan-pinto replied:
latest compatible

ilan-pinto commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+RUN chmod +x /opt/keycloak/bin/user-import.sh
+RUN chmod +x /opt/keycloak/bin/user-export.sh
+RUN chmod +x /opt/keycloak/bin/groups-ids-wrapper.sh
+
```
Every line in the docker file adds a layer in the docker image, increasing the size. 
https://docs.docker.com/build/guide/layers/ 
e.g try to cluster the copy commands

nemerna replied:
Sure

ilan-pinto commented on Entrypoint.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
```
That's the main file and the only one visible in the root folder - rename it to keyclock-user-cli or similar

ilan-pinto commented on Entrypoint.sh:
```diff
@@ -0,0 +1,17 @@
+elif [ "$1" = "export" ]; then
+    echo "Performing export..."
+    export WORK_DIRECTORY=/home/default/EXPORT_DIR
+    user-export.sh
```
Move user-export.sh and  user-import.sh and accordingly change path 

also i would add an "help" option

nemerna replied:
i didnt understand
`
Move user-export.sh and user-import.sh and accordingly change path

`

ilan-pinto replied:
Goes with the previous comment 
user-export. sh and user-import. sh should not be in the root folder as you don't expect users to use them directly

nemerna replied:
you mean in the repo?
because the code you commented about is relevant only inside the container,

ilan-pinto commented on README.md:
Is there an example of the user.csv file?

nemerna replied:
yep, it is a json not a csv
-  below is the username/id for all server users
[ {
  "id" : "c29a1789-fb7f-4e11-a0a0-61f2d30fe009",
  "username" : "kcm"
}, {
  "id" : "1a468df5-1ab1-4144-9d8c-736dac74181b",
  "username" : "names"
}, {
  "id" : "31d4e96a-c982-450b-9dec-d6639c23d65a",
  "username" : "namess"
} ]

- this is an example of a single user
{
  "username" : "kcm",
  "enabled" : true,
  "totp" : false,
  "emailVerified" : true,
  "firstName" : "KCM",
  "lastName" : "Admin",
  "email" : "temp@test.com",
  "attributes" : {
    "test" : [ "test" ]
  },
  "disableableCredentialTypes" : [ ],
  "requiredActions" : [ ],
  "notBefore" : 0,
  "access" : {
    "manageGroupMembership" : true,
    "view" : true,
    "mapRoles" : true,
    "impersonate" : true,
    "manage" : true
  }
}

ilan-pinto commented on user-export.sh:
```diff
@@ -1,10 +1,29 @@
 #!/bin/bash
 
-echo "starting exporter"
+export ALL_USERS_FILE=${WORK_DIRECTORY}/EXPORT_SERVER_USERS.json
```
why ALL_USERS_FILE and not just USERS_FILE?

nemerna replied:
because,
in a step i get the entire user ids&usernames in a single get into the file ${WORK_DIRECTORY}/EXPORT_SERVER_USERS.json

then for each user i make separate get

the reason for that, is when you try to get all users, you do not get the entire related user information, but only the tope level ones (user Attributes cant be retrieved using "get users", instead "get users/$UID)
similar for groups (get users/$UID/groups)

ilan-pinto commented on user-export.sh:
```diff
@@ -25,7 +44,7 @@ do
+  kcadm.sh get users/$user_id --fields '*(*(*(*(*(*))))),-id,-createdTimestamp' --realm $EXPORT_REALM --server $EXPORT_KEYCLOAK_SERVER --token $EXPORT_TOKEN --no-config > $WORK_DIRECTORY/$username/USER.json
 
   # If failed to get user, skip the current iteration
   if [ $? -ne 0 ]; then
```
Consider writing to a log file

nemerna replied:
sure, will make it through the Makefile 
(container output redirect), where ir un the container in interactive mode

ilan-pinto commented on client/lib/bcprov-jdk15on-1.70.jar:
What is this file?

nemerna replied:
its again based on Danielles Dockerfile, and the lib dir is from the official Keycloak release,
but tried to remove the library and nothing got broken ...

nemerna committed:
reduce number of actions in Dockerfile
, move scripts to scripts directory, remove all the keycloak libs/modules, as we will be geting them always from the image

nemerna committed:
made the docker un interactive, for supporting output redirect to logfile
